syntax = "proto3";

package model;

option go_package = "github.com/17882237881/MaaS/shared/proto/model;modelpb";

import "google/protobuf/timestamp.proto";
import "google/protobuf/empty.proto";

// ModelService provides CRUD operations for ML models
service ModelService {
  // Create a new model
  rpc CreateModel(CreateModelRequest) returns (CreateModelResponse);
  
  // Get model by ID
  rpc GetModel(GetModelRequest) returns (GetModelResponse);
  
  // List models with pagination and filtering
  rpc ListModels(ListModelsRequest) returns (ListModelsResponse);
  
  // Update model information
  rpc UpdateModel(UpdateModelRequest) returns (UpdateModelResponse);
  
  // Delete model by ID
  rpc DeleteModel(DeleteModelRequest) returns (google.protobuf.Empty);
  
  // Update model status
  rpc UpdateModelStatus(UpdateModelStatusRequest) returns (UpdateModelStatusResponse);
  
  // Add tags to model
  rpc AddModelTags(AddModelTagsRequest) returns (google.protobuf.Empty);
  
  // Remove tags from model
  rpc RemoveModelTags(RemoveModelTagsRequest) returns (google.protobuf.Empty);
  
  // Set model metadata
  rpc SetModelMetadata(SetModelMetadataRequest) returns (google.protobuf.Empty);
  
  // Get model metadata
  rpc GetModelMetadata(GetModelMetadataRequest) returns (GetModelMetadataResponse);
}

// Model represents a machine learning model
message Model {
  string id = 1;
  string name = 2;
  string description = 3;
  string version = 4;
  string framework = 5;
  string status = 6;
  int64 size = 7;
  string checksum = 8;
  string storage_path = 9;
  string docker_image = 10;
  repeated string tags = 11;
  string owner_id = 12;
  string tenant_id = 13;
  bool is_public = 14;
  google.protobuf.Timestamp created_at = 15;
  google.protobuf.Timestamp updated_at = 16;
}

// CreateModelRequest is the request for CreateModel
message CreateModelRequest {
  string name = 1;
  string description = 2;
  string version = 3;
  string framework = 4;
  repeated string tags = 5;
  map<string, string> metadata = 6;
  string owner_id = 7;
  string tenant_id = 8;
  bool is_public = 9;
}

// CreateModelResponse is the response for CreateModel
message CreateModelResponse {
  Model model = 1;
}

// GetModelRequest is the request for GetModel
message GetModelRequest {
  string id = 1;
}

// GetModelResponse is the response for GetModel
message GetModelResponse {
  Model model = 1;
}

// ListModelsRequest is the request for ListModels
message ListModelsRequest {
  string name = 1;
  string framework = 2;
  string status = 3;
  string owner_id = 4;
  string tenant_id = 5;
  repeated string tags = 6;
  bool is_public = 7;
  int32 page = 8;
  int32 limit = 9;
}

// ListModelsResponse is the response for ListModels
message ListModelsResponse {
  repeated Model models = 1;
  int64 total = 2;
  int32 page = 3;
  int32 limit = 4;
}

// UpdateModelRequest is the request for UpdateModel
message UpdateModelRequest {
  string id = 1;
  string name = 2;
  string description = 3;
  repeated string tags = 4;
  map<string, string> metadata = 5;
  bool is_public = 6;
}

// UpdateModelResponse is the response for UpdateModel
message UpdateModelResponse {
  Model model = 1;
}

// DeleteModelRequest is the request for DeleteModel
message DeleteModelRequest {
  string id = 1;
}

// UpdateModelStatusRequest is the request for UpdateModelStatus
message UpdateModelStatusRequest {
  string id = 1;
  string status = 2;
}

// UpdateModelStatusResponse is the response for UpdateModelStatus
message UpdateModelStatusResponse {
  Model model = 1;
}

// AddModelTagsRequest is the request for AddModelTags
message AddModelTagsRequest {
  string model_id = 1;
  repeated string tags = 2;
}

// RemoveModelTagsRequest is the request for RemoveModelTags
message RemoveModelTagsRequest {
  string model_id = 1;
  repeated string tags = 2;
}

// SetModelMetadataRequest is the request for SetModelMetadata
message SetModelMetadataRequest {
  string model_id = 1;
  map<string, string> metadata = 2;
}

// GetModelMetadataRequest is the request for GetModelMetadata
message GetModelMetadataRequest {
  string model_id = 1;
}

// GetModelMetadataResponse is the response for GetModelMetadata
message GetModelMetadataResponse {
  map<string, string> metadata = 1;
}
